src/clients/c++/experimental_api_v2/library/common.h

```diff
@@ -0,0 +1,561 @@
+  /// the size expected by the input.
+  /// \return Error object indicating success or failure.
+  Error SetRaw(const uint8_t* input, size_t input_byte_size);
+
+  /// Set tensor values for this input by reference into a shared memory
+  /// region. The values are not copied and so the shared memory region and
+  /// its contents must not be modified or destroyed until this input is no
+  /// longer needed (that is until the Run() call(s) that use the input have
+  /// completed. This function must be called a single time for an input that
+  /// is using shared memory. For batched inputs, the tensor values for the
+  /// entire batch must be contiguous in a single shared memory region.
+  /// Note: The options must be set using SetRunOptions before calling the
+  /// SetSharedMemory function since the batch size is needed for validation.
+  /// \param name The user-given name for the registered shared memory region
+  /// where the tensor values for this input is stored.
+  /// \param byte_size The size, in bytes of the input tensor data. Must
+  /// match the size expected for the input shape.
+  /// \param offset The offset into the shared memory region upto the start
+  /// of the input tensor values. The default value is 0.
+  /// \return Error object indicating success or failure
+  Error SetSharedMemory(
+      const std::string& name, size_t byte_size, size_t offset = 0);
+
+  /// Set tensor values for this input from a vector or
+  /// strings. This method can only be used for tensors with STRING
+  /// data-type. The strings are assigned in row-major order to the
+  /// elements of the tensor. The strings are copied and so the
+  /// 'input' does not need to be preserved as with SetRaw(). For
+  /// batched inputs this function must be called batch-size times
+  /// to provide all tensor values for a batch of this input.
+  /// \param input The vector holding tensor string values.
+  /// \return Error object indicating success or failure.
+  Error SetFromString(const std::vector<std::string>& input);
+
+  /// Copy the shared memory region name, batch_byte_size and offset.
+  /// \param name The user-given name for the registered shared memory region
+  /// where the tensor values for this input is stored.
+  /// \param byte_size The size, in bytes of the input tensor data. Must
+  /// match the size expected for the input batch.
+  /// \param offset The offset into the shared memory region upto the start
+  /// of the input tensor values.
+  /// \return Error object indicating success or failure
+  Error SharedMemoryInfo(
+      std::string* name, size_t* batch_byte_size, size_t* offset) const;
+
+  // Prepare to send this input as part of a request.
+  Error PrepareForRequest();
+
+  // Copy into 'buf' up to 'size' bytes of this input's data. Return
+  // the actual amount copied in 'input_bytes' and if the end of input
+  // is reached in 'end_of_input'
+  Error GetNext(
```

Comment:
Why do we need to provide public methods to get the data out of an input? The user put the data in so they should already have it.